src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs

```diff
@@ -19,28 +19,28 @@ public RewriteAction(RuleTerminiation result, Pattern pattern, bool clearQuery)
             ClearQuery = clearQuery;
         }
 
-        public override RuleResult ApplyAction(HttpContext context, MatchResults ruleMatch, MatchResults condMatch)
+        public override RuleResult ApplyAction(RewriteContext context, MatchResults ruleMatch, MatchResults condMatch)
         {
             var pattern = Url.Evaluate(context, ruleMatch, condMatch);
 
             if (ClearQuery)
             {
-                context.Request.QueryString = new QueryString();
+                context.HttpContext.Request.QueryString = new QueryString();
             }
             // TODO PERF, substrings, object creation, etc.
             if (pattern.IndexOf("://") >= 0)
             {
                 string scheme = null;
-                var host = new HostString();
-                var path = new PathString();
-                var query = new QueryString();
-                var fragment = new FragmentString();
+                HostString host;
+                PathString path;
+                QueryString query;
+                FragmentString fragment;
                 UriHelper.FromAbsolute(pattern, out scheme, out host, out path, out query, out fragment);
 
-                context.Request.Scheme = scheme;
-                context.Request.Host = host;
-                context.Request.Path = path;
-                context.Request.QueryString = query.Add(context.Request.QueryString);
+                context.HttpContext.Request.Scheme = scheme;
+                context.HttpContext.Request.Host = host;
+                context.HttpContext.Request.Path = path;
```

Comment:
https://github.com/aspnet/BasicMiddleware/issues/59, filed and will be handled in another pr.